More early resolve issues

jedwards4b opened the pull request:
This change unresolves some variables in the case that were being resolved too early, this resulted in some of the perl code breaking which I've put a bandaid on in the hope that it will be replaced soon.  

Test suite: scripts_regression_tests
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes  #406

User interface changes?: 

Code review:

jedwards4b commented on cime_config/cesm/machines/config_batch.xml:
```diff
@@ -424,7 +424,8 @@
   <batch_system MACH="yellowstone" type="lsf" version="9.1">
     <queues>
-      <queue walltimemax="24:00" jobmin="1" jobmax="8">caldera</queue>
+      <queue walltimemax="24:00" jobmin="1" jobmax="1" jobname="case.lt_archive">hpss</queue>
```
Add a jobname attribute to the queue node.  This allows me to force the case.lt_archive script to use the hpss queue on yellowstone.

jedwards4b commented on components/data_comps/dice/bld/build-namelist:
```diff
@@ -306,6 +306,9 @@ if (defined $opts{'infile'}) {
 my %xmlvars = ();
 SetupTools::getxmlvars(${CASEROOT},\%xmlvars);
+# need to expand DIN_LOC_ROOT first
+$xmlvars{DIN_LOC_ROOT}=SetupTools::expand_xml_var($xmlvars{DIN_LOC_ROOT}, \%xmlvars);
```
If this variable isn't resolved first the result in %xmlvars is corrupted.   This is the bandaid.

jedwards4b commented on driver_cpl/cime_config/config_component.xml:
```diff
@@ -1146,14 +1146,6 @@
     <desc>number of wav cells in j direction - DO NOT EDIT (for experts only)</desc>
   </entry>
 
-  <entry id="MASK_GRID">
```
Never used anyplace.  Available from the grid longname.

jedwards4b commented on driver_cpl/cime_config/config_component.xml:
```diff
@@ -1766,14 +1758,6 @@
     <desc>Machine name</desc>
   </entry>
 
-  <entry id="MACHINES_FILE">
```
Not used.

jgfouca replied:
Wasn't this fixed in a previous PR?

jedwards4b commented on scripts/create_newcase:
```diff
@@ -154,9 +154,9 @@ def _main_func():
-        case.set_value("CASE", os.path.basename(caseroot))
-        case.set_value("CASEROOT", caseroot)
-        case.set_value("SRCROOT", srcroot)
+        case.set_lookup_value("CASE", os.path.basename(caseroot))
```
Fixes issue 389

jedwards4b commented on utils/perl5lib/Streams/TemplateGeneric.pm:
```diff
@@ -463,8 +465,8 @@ sub GetDataFilepath {
   if ( ! defined($self->{'template'}) ) {
      die "${nm}:: a template has NOT been read in yet -- abort.\n";
   }
-  my $defaults_ref = $self->{'defaults'};
```
I couldn't help myself and tidied up this file in a couple of places removing unused variables.

jgfouca replied:
Just curious since this is Perl, which tool invokes this code?

jedwards4b replied:
build-namelist

rljacob commented:
Does this fix #404 or just address part of it?

jedwards4b commented on utils/python/CIME/XML/env_batch.py:
```diff
@@ -330,6 +330,8 @@ def get_submit_args(self, case, job):
             if name is None:
                 submitargs+=" %s"%flag
             else:
+                if name.startswith("$"):
```
In config_batch.xml the submit_args node references other xml variables, we need to strip the $ from the front of the variable so that we can resolve it. We have to do this explicitly because this is an attribute not a node,

jedwards4b commented on utils/python/CIME/XML/env_batch.py:
```diff
@@ -473,8 +475,12 @@ def select_best_queue(self, num_pes):
             if queue is not None:
                 jobmin = queue.get("jobmin")
                 jobmax = queue.get("jobmax")
+                jobname = queue.get("jobname")
```
This implements the jobname attribute of queue.

jedwards4b commented on utils/python/CIME/buildnml.py:
```diff
@@ -154,7 +154,9 @@ def _build_data_nml(case, caseroot, compclass):
         rc, out, err = run_cmd(cmd, from_dir=confdir)
         expect(rc==0,"Command %s failed rc=%d\nout=%s\nerr=%s"%(cmd,rc,out,err))
-
+        if out is not None:
```
useful for debugging.  Not too noisy.

jedwards4b commented on utils/python/CIME/case.py:
```diff
@@ -80,18 +80,16 @@ def __init__(self, case_root=None, read_only=True):
+        # table and then remove the entry.
         self.lookups = {}
-        self.lookups['CIMEROOT'] = os.path.abspath(get_cime_root())
+        self.set_lookup_value('CIMEROOT',os.path.abspath(get_cime_root()))
```
try to use the method consistently

jedwards4b commented on utils/python/CIME/case.py:
```diff
@@ -80,18 +80,16 @@ def __init__(self, case_root=None, read_only=True):
         self._pesfile = None
         self._gridfile = None
         self._components = []
-        self._component_config_files = []
```
not needed

jedwards4b commented on utils/python/CIME/case.py:
```diff
@@ -384,6 +383,10 @@ def _get_component_config_data(self):
         # Determine list of component classes that this coupler/driver knows how
         # to deal with. This list follows the same order as compset longnames follow.
         files = Files()
+        # Add the group and elements for the config_files.xml
```
This needs to be done before setting values.

jedwards4b commented:
I believe that it fixes 406

jgfouca commented:
This all looks fine. I could swear some of these exact same changes were merged yesterday though, so that confuses me.

jedwards4b commented:
They were merged Monday - but I found several failing tests related to that perl build-namelist so I reverted the merge and fixed those issues before issuing the PR again.  Sorry for the confusion.

jgfouca commented:
Ah, forgot about that, thanks.
